### PR TITLE
Improve `head` injection behavior

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- Improves Astro's head injection behavior per [RFC0007](https://github.com/withastro/rfcs/blob/main/proposals/0007-finalize-head-body.md)
- Relies on https://github.com/withastro/compiler/pull/315

## Testing

Existing tests updated

## Docs

N/A

### CODE REVIEW

**Comment by FredKSchott** [comment]
Closing as stale, feel free to reopen when ready!

**Review by FredKSchott — APPROVED** [review]

File: packages/astro/src/runtime/server/index.ts [hunk 1]
```diff
@@ -433,15 +419,34 @@ export async function renderEndpoint(mod: EndpointHandler, params: any) {
...
+	const children = item.children;
+	return index === all.findIndex((i) => JSON.stringify(i.props) === props && i.children == children);
+};
+
+
 // Renders a page to completion by first calling the factory callback, waiting for its result, and then appending
 // styles and scripts into the head.
-export async function renderPage(result: SSRResult, Component: AstroComponentFactory, props: any, children: any) {
-	const template = await renderToString(result, Component, props, children);
+export async function renderHead(result: SSRResult) {
```
Comment (natemoo-re) [inline packages/astro/src/runtime/server/index.ts:445]: No more `renderPage` entrypoint, just use `renderToString`.

File: packages/astro/src/runtime/server/index.ts [hunk 2]
```diff
@@ -433,15 +419,34 @@ export async function renderEndpoint(mod: EndpointHandler, params: any) {
...
+// Calls a component and renders it into a string of HTML
+export async function renderToString(result: SSRResult, componentFactory: AstroComponentFactory, props: any, children: any) {
+	const Component = await componentFactory(result, props, children);
+	let template = await renderAstroComponent(Component);
+
+	// <!--astro:head--> injected by compiler
+	// Must be handled at the end of the rendering process
+	if (template.indexOf('<!--astro:head-->') > -1) {
+		template = template.replace('<!--astro:head-->', await renderHead(result));
+	}
```
Comment (natemoo-re) [inline packages/astro/src/runtime/server/index.ts:431]: `<!--astro:head-->` is injected by the compiler. If we see this, it's at the end of the `<head>` element and we should inject the links/styles/scripts.

File: packages/astro/src/runtime/server/index.ts [hunk 3]
```diff
@@ -462,17 +467,10 @@ export async function renderPage(result: SSRResult, Component: AstroComponentFac
...
 		.filter(uniqueElements)
 		.map((link) => renderElement('link', link, false));
-
-	// inject styles & scripts at end of <head>
-	let headPos = template.indexOf('</head>');
-	if (headPos === -1) {
-		return links.join('\n') + styles.join('\n') + scripts.join('\n') + template; // if no </head>, prepend styles & scripts
-	}
-	return template.substring(0, headPos) + links.join('\n') + styles.join('\n') + scripts.join('\n') + template.substring(headPos);
+	return unescapeHTML(links.join('\n') + styles.join('\n') + scripts.join('\n') + '\n' + '<!--astro:head:injected-->');
```
Comment (natemoo-re) [inline packages/astro/src/runtime/server/index.ts:473]: We also add a `<!--astro:head:injected-->` comment to flag that we've already injected the links/styles/scripts.

File: packages/astro/src/vite-plugin-astro/compile.ts [hunk 4]
```diff
@@ -47,7 +44,6 @@ async function compile(config: AstroConfig, filename: string, source: string, vi
 	// use `sourcemap: "both"` so that sourcemap is included in the code
 	// result passed to esbuild, but also available in the catch handler.
 	const transformResult = await transform(source, {
-		as: isPage ? 'document' : 'fragment',
```
Comment (natemoo-re) [inline packages/astro/src/vite-plugin-astro/compile.ts:50]: No more `as: 'document'|'fragment'`! We no longer care, every Astro file is parsed the same.

File: packages/astro/src/core/render/core.ts [hunk 5]
```diff
@@ -97,7 +97,14 @@ export async function render(opts: RenderOptions): Promise<string> {
 		scripts,
 	});
 
-	let html = await renderPage(result, Component, pageProps, null);
+	let html = await renderToString(result, Component, pageProps, null);
+
+	// handle final head injection if it hasn't happened already
+	if (html.indexOf("<!--astro:head:injected-->") == -1) {
+		html = await renderHead(result) + html;
+	}
```
Comment (natemoo-re) [inline packages/astro/src/core/render/core.ts:105]: Slight deviation from [RFC0007](https://github.com/withastro/rfcs/blob/main/proposals/0007-finalize-head-body.md)... if we haven't injected any head elements yet, the user may not have rendered a `<head>` element so we just prepend them to the file.

File: packages/astro/src/core/render/core.ts [hunk 6]
```diff
@@ -97,7 +97,14 @@ export async function render(opts: RenderOptions): Promise<string> {
...
+	// cleanup internal state flags
+	html = html.replace("<!--astro:head:injected-->", '');
```
Comment (natemoo-re) [inline packages/astro/src/core/render/core.ts:107]: Remove our flag for a nicer output.

File: packages/astro/src/runtime/server/index.ts (hunk 3 above)
Comment (matthewp) [reply packages/astro/src/runtime/server/index.ts:473]: Why?

Comment (matthewp) [reply packages/astro/src/runtime/server/index.ts:473]: I think I understand after reading the code below.

**Comment by matthewp** [comment]
lgtm

**Pull request merged**